File: client-scalaz/src/main/scala/org/openapitools/client/api/Retailer.scala
package org.openapitools.client.api

import argonaut._
import argonaut.EncodeJson._
import argonaut.DecodeJson._

import org.http4s.{EntityDecoder, EntityEncoder}
import org.http4s.argonaut._
import org.joda.time.DateTime


import Retailer._

case class Retailer (
  id: Option[Long],
active: Option[Boolean],
valid: Option[Boolean],
latitude: Option[Double],
longitude: Option[Double],
locationDescription: Option[String],
name: Option[String],
responsible: Option[Account],
manager: Option[Account],
detailsHeader: Option[String],
detailsBody: Option[String],
detailsLineLeft1: Option[String],
detailsLineLeft2: Option[String],
detailsLineRight1: Option[String],
metaDescription: Option[String],
directions: Option[String],
hours: Option[String],
homePhone: Option[String],
cellPhone: Option[String],
cellCarrier: Option[CellCarrier],
businessPhone: Option[String],
businessPhoneExt: Option[String],
faxNumber: Option[String],
timeZone: Option[String],
utcOffset: Option[String],
code501c3: Option[String],
emailAddress: Option[String],
address: Option[Address],
web: Option[String],
featured: Option[Boolean],
listed: Option[Boolean],
categories: Option[List[Category]],
territory: Option[Territory],
locations: Option[List[RetailerLocation]],
yipitId: Option[Long],
facebookUrl: Option[String],
twitterUrl: Option[String],
visibility: Option[Visibility],
notes: Option[List[Note]],
noteCount: Option[Long],
appKey: Option[String],
categoryTree: Option[String],
filterTree: Option[String],
billableId: Option[Long],
subType: Option[String],
contentName: Option[String],
owner: Option[Account])

object Retailer {
  import DateTimeCodecs._
  sealed trait Visibility
  case object PUBLIC extends Visibility
  case object `PRIVATE` extends Visibility
  case object FRIENDS extends Visibility

  object Visibility {
    def toVisibility(s: String): Option[Visibility] = s match {
      case "PUBLIC" => Some(PUBLIC)
      case "`PRIVATE`" => Some(`PRIVATE`)
      case "FRIENDS" => Some(FRIENDS)
      case _ => None
    }

    def fromVisibility(x: Visibility): String = x match {
      case PUBLIC => "PUBLIC"
      case `PRIVATE` => "`PRIVATE`"
      case FRIENDS => "FRIENDS"
    }
  }

  implicit val VisibilityEnumEncoder: EncodeJson[Visibility] =
    EncodeJson[Visibility](is => StringEncodeJson(Visibility.fromVisibility(is)))

  implicit val VisibilityEnumDecoder: DecodeJson[Visibility] =
    DecodeJson.optionDecoder[Visibility](n => n.string.flatMap(jStr => Visibility.toVisibility(jStr)), "Visibility failed to de-serialize")

  implicit val RetailerCodecJson: CodecJson[Retailer] = CodecJson.derive[Retailer]
  implicit val RetailerDecoder: EntityDecoder[Retailer] = jsonOf[Retailer]
  implicit val RetailerEncoder: EntityEncoder[Retailer] = jsonEncoderOf[Retailer]
}

File: client-clojure/src/sirqul_io_t_platform/specs/retailer.clj
(ns sirqul-io-t-platform.specs.retailer
  (:require [clojure.spec.alpha :as s]
            [spec-tools.data-spec :as ds]
            [sirqul-io-t-platform.specs.account :refer :all]
            [sirqul-io-t-platform.specs.account :refer :all]
            [sirqul-io-t-platform.specs.cell-carrier :refer :all]
            [sirqul-io-t-platform.specs.address :refer :all]
            [sirqul-io-t-platform.specs.category :refer :all]
            [sirqul-io-t-platform.specs.territory :refer :all]
            [sirqul-io-t-platform.specs.retailer-location :refer :all]
            [sirqul-io-t-platform.specs.note :refer :all]
            [sirqul-io-t-platform.specs.account :refer :all]
            )
  (:import (java.io File)))


(def retailer-data
  {
   (ds/opt :id) int?
   (ds/opt :active) boolean?
   (ds/opt :valid) boolean?
   (ds/opt :latitude) float?
   (ds/opt :longitude) float?
   (ds/opt :locationDescription) string?
   (ds/opt :name) string?
   (ds/opt :responsible) account-spec
   (ds/opt :manager) account-spec
   (ds/opt :detailsHeader) string?
   (ds/opt :detailsBody) string?
   (ds/opt :detailsLineLeft1) string?
   (ds/opt :detailsLineLeft2) string?
   (ds/opt :detailsLineRight1) string?
   (ds/opt :metaDescription) string?
   (ds/opt :directions) string?
   (ds/opt :hours) string?
   (ds/opt :homePhone) string?
   (ds/opt :cellPhone) string?
   (ds/opt :cellCarrier) cell-carrier-spec
   (ds/opt :businessPhone) string?
   (ds/opt :businessPhoneExt) string?
   (ds/opt :faxNumber) string?
   (ds/opt :timeZone) string?
   (ds/opt :utcOffset) string?
   (ds/opt :code501c3) string?
   (ds/opt :emailAddress) string?
   (ds/opt :address) address-spec
   (ds/opt :web) string?
   (ds/opt :featured) boolean?
   (ds/opt :listed) boolean?
   (ds/opt :categories) (s/coll-of category-spec)
   (ds/opt :territory) territory-spec
   (ds/opt :locations) (s/coll-of retailer-location-spec)
   (ds/opt :yipitId) int?
   (ds/opt :facebookUrl) string?
   (ds/opt :twitterUrl) string?
   (ds/opt :visibility) string?
   (ds/opt :notes) (s/coll-of note-spec)
   (ds/opt :noteCount) int?
   (ds/opt :appKey) string?
   (ds/opt :categoryTree) string?
   (ds/opt :filterTree) string?
   (ds/opt :billableId) int?
   (ds/opt :subType) string?
   (ds/opt :contentName) string?
   (ds/opt :owner) account-spec
   })

(def retailer-spec
  (ds/spec
    {:name ::retailer
     :spec retailer-data}))
